src/Octoshift/Services/ArchiveUploader.cs

```diff
@@ -0,0 +1,167 @@
+            {
+                nextUrl = await UploadPart(buffer, bytesRead, nextUrl.ToString(), partsRead, totalParts);
+                partsRead++;
+            }
+
+            // 3. Complete the upload
+            await CompleteUpload(nextUrl.ToString());
+
+            return $"gei://archive/{guid}";
+        }
+        catch (Exception ex)
+        {
+            throw new OctoshiftCliException("Failed during multipart upload.", ex);
+        }
+    }
+
+    private async Task<IEnumerable<KeyValuePair<string, IEnumerable<string>>>> StartUpload(string uploadUrl, string archiveName, long contentSize)
+    {
+        _log.LogInformation($"Starting archive upload into GitHub owned storage: {archiveName}...");
+
+        var body = new
+        {
+            content_type = "application/octet-stream",
+            name = archiveName,
+            size = contentSize
+        };
+
+        try
+        {
+            var response = await _client.PostWithFullResponseAsync(uploadUrl, body);
+            return response.ResponseHeaders.ToList();
+        }
+        catch (Exception ex)
+        {
+            throw new OctoshiftCliException("Failed to start upload.", ex);
+        }
+    }
+
+    private async Task<Uri> UploadPart(byte[] body, int bytesRead, string nextUrl, int partsRead, long totalParts)
+    {
+        _log.LogInformation($"Uploading part {partsRead + 1}/{totalParts}...");
+        using var content = new ByteArrayContent(body, 0, bytesRead);
+        content.Headers.ContentType = new("application/octet-stream");
+
+        try
+        {
+            // Make the PATCH request and retrieve headers
+            var patchResponse = await _client.PatchWithFullResponseAsync(nextUrl, content);
+            var headers = patchResponse.ResponseHeaders.ToList();
+
+            // Retrieve the next URL from the response headers
+            return GetNextUrl(headers);
+        }
+        catch (Exception ex)
+        {
+            throw new OctoshiftCliException("Failed to upload part.", ex);
+        }
+    }
+
+    private async Task CompleteUpload(string lastUrl)
+    {
+        try
+        {
+            await _client.PutAsync(lastUrl, "");
+            _log.LogInformation("Finished uploading archive");
+        }
+        catch (Exception ex)
+        {
+            throw new OctoshiftCliException("Failed to complete upload.", ex);
+        }
+    }
+
+    private Uri GetNextUrl(IEnumerable<KeyValuePair<string, IEnumerable<string>>> headers)
+    {
+        // Use FirstOrDefault to safely handle missing Location headers
+        var locationHeader = headers.First(header => header.Key.Equals("Location", StringComparison.OrdinalIgnoreCase));
+
+        if (!string.IsNullOrEmpty(locationHeader.Key))
+        {
+            var locationValue = locationHeader.Value.FirstOrDefault();
+            if (locationValue.HasValue())
+            {
+                return new Uri(new Uri(BASE_URL), locationValue);
```

Comment:
There is a bug here that I discovered when fiddling with the unit tests. The base url is `https://uploads.github.com/organizations` but it's actually not a base url but rather a full url. The base here would be `https://uploads.github.com`. So if you call `.ToString()` on the Uri object that is getting returned from the method it will actually omit the `organizations`:

![image](https://github.com/user-attachments/assets/37d13195-99f8-40bf-8ef5-9ca6d6039f98)

To fix we can compose the url like below or alternatively we can change the base url and prepend `organizations` to the `locationValue`, I guess the first is just easier:

```c#
new Uri($"{BASE_URL}/{locationValue}")
```

Reply:
fixed in c93e44df9a493fb9c0b2e91feab03e471aefa368

Reply:
Fixed this to account for the bug within the 🐛 [ede2cd8](https://github.com/github/gh-gei/pull/1285/commits/ede2cd8616b8c43ccc95b730f31e7f976327054e)